File: CRMAdapter/VastOnline/Adapter/InvoiceAdapter.cs
/*
 * File: InvoiceAdapter.cs
 * Role: Implements canonical invoice access for the Vast Online backend.
 * Architectural Purpose: Converts Azure SQL invoice data into the CRM canonical invoice aggregate with line items.
 */
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using CRMAdapter.CommonConfig;
using CRMAdapter.CommonContracts;
using CRMAdapter.CommonDomain;

namespace CRMAdapter.VastOnline.Adapter
{
    /// <summary>
    /// Vast Online implementation of the <see cref="IInvoiceAdapter"/> contract.
    /// </summary>
    public sealed class InvoiceAdapter : SqlAdapterBase, IInvoiceAdapter
    {
        private const int DefaultListLimit = 100;

        private static readonly string[] RequiredInvoiceKeys =
        {
            "Invoice.__source",
            "Invoice.Id",
            "Invoice.CustomerId",
            "Invoice.VehicleId",
            "Invoice.Number",
            "Invoice.Date",
            "Invoice.Total",
            "Invoice.Status"
        };

        private static readonly string[] RequiredInvoiceLineKeys =
        {
            "InvoiceLine.__source",
            "InvoiceLine.InvoiceId",
            "InvoiceLine.Description",
            "InvoiceLine.Quantity",
            "InvoiceLine.UnitPrice",
            "InvoiceLine.Tax"
        };

        private static readonly string[] InvoiceProjectionFields =
        {
            "Id",
            "CustomerId",
            "VehicleId",
            "Number",
            "Date",
            "Total",
            "Status"
        };

        private readonly int _defaultListLimit;

        /// <summary>
        /// Initializes a new instance of the <see cref="InvoiceAdapter"/> class.
        /// </summary>
        /// <param name="connection">Database connection.</param>
        /// <param name="fieldMap">Field map configuration.</param>
        /// <param name="defaultListLimit">Default maximum results for list operations.</param>
        public InvoiceAdapter(DbConnection connection, FieldMap fieldMap, int defaultListLimit = DefaultListLimit)
            : base(connection, fieldMap)
        {
            _defaultListLimit = defaultListLimit;
            MappingValidator.EnsureMappings(fieldMap, RequiredInvoiceKeys, nameof(InvoiceAdapter));
            MappingValidator.EnsureMappings(fieldMap, RequiredInvoiceLineKeys, nameof(InvoiceAdapter));
        }

        /// <inheritdoc />
        public async Task<Invoice?> GetByIdAsync(Guid id, CancellationToken cancellationToken = default)
        {
            var fieldMap = FieldMap.GetTargets("Invoice", InvoiceProjectionFields);
            var source = FieldMap.GetEntitySource("Invoice");
            var selectClause = string.Join(", ", fieldMap.Select(kvp => $"{kvp.Value} AS [{kvp.Key}]"));
            var commandText = $"SELECT {selectClause} FROM {source} WHERE {fieldMap["Id"]} = @id";

            await using var command = await CreateCommandAsync(commandText, cancellationToken).ConfigureAwait(false);
            AddParameter(command, "@id", id);

            await using var reader = await command.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);
            if (!await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
            {
                return null;
            }

            var invoice = ReadInvoice(reader);
            var lines = await LoadInvoiceLinesAsync(new[] { invoice.Id }, cancellationToken).ConfigureAwait(false);
            var lineItems = lines.TryGetValue(invoice.Id, out var collection)
                ? collection
                : Array.Empty<InvoiceLine>();

            return new Invoice(
                invoice.Id,
                invoice.CustomerId,
                invoice.VehicleId,
                invoice.InvoiceNumber,
                invoice.InvoiceDate,
                invoice.TotalAmount,
                invoice.Status,
                lineItems);
        }

        /// <inheritdoc />
        public async Task<IReadOnlyCollection<Invoice>> GetByCustomerAsync(
            Guid customerId,
            int maxResults,
            CancellationToken cancellationToken = default)
        {
            var limit = Math.Min(_defaultListLimit, maxResults > 0 ? maxResults : _defaultListLimit);
            var fieldMap = FieldMap.GetTargets("Invoice", InvoiceProjectionFields);
            var source = FieldMap.GetEntitySource("Invoice");
            var selectClause = string.Join(", ", fieldMap.Select(kvp => $"{kvp.Value} AS [{kvp.Key}]"));
            var commandText = $@"SELECT TOP (@limit) {selectClause}
FROM {source}
WHERE {fieldMap["CustomerId"]} = @customerId
ORDER BY {fieldMap["Date"]} DESC;";

            await using var command = await CreateCommandAsync(commandText, cancellationToken).ConfigureAwait(false);
            AddParameter(command, "@limit", limit, DbType.Int32);
            AddParameter(command, "@customerId", customerId);

            var invoices = new List<InvoiceRecord>();
            await using var reader = await command.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);
            while (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
            {
                invoices.Add(ReadInvoice(reader));
            }

            var lineLookup = await LoadInvoiceLinesAsync(invoices.Select(i => i.Id), cancellationToken).ConfigureAwait(false);
            var results = invoices.Select(record =>
            {
                var lines = lineLookup.TryGetValue(record.Id, out var items)
                    ? items
                    : Array.Empty<InvoiceLine>();
                return new Invoice(
                    record.Id,
                    record.CustomerId,
                    record.VehicleId,
                    record.InvoiceNumber,
                    record.InvoiceDate,
                    record.TotalAmount,
                    record.Status,
                    lines);
            }).ToList();

            return new ReadOnlyCollection<Invoice>(results);
        }

        private async Task<IReadOnlyDictionary<Guid, IReadOnlyCollection<InvoiceLine>>> LoadInvoiceLinesAsync(
            IEnumerable<Guid> invoiceIds,
            CancellationToken cancellationToken)
        {
            var ids = invoiceIds.Distinct().ToList();
            if (ids.Count == 0)
            {
                return new Dictionary<Guid, IReadOnlyCollection<InvoiceLine>>();
            }

            var fields = FieldMap.GetTargets("InvoiceLine", new[] { "InvoiceId", "Description", "Quantity", "UnitPrice", "Tax" });
            var source = FieldMap.GetEntitySource("InvoiceLine");
            var parameterNames = ids.Select((_, index) => $"@i{index}").ToArray();
            var selectClause = $"{fields["InvoiceId"]} AS [InvoiceId], {fields["Description"]} AS [Description], {fields["Quantity"]} AS [Quantity], {fields["UnitPrice"]} AS [UnitPrice], {fields["Tax"]} AS [Tax]";
            var commandText = $"SELECT {selectClause} FROM {source} WHERE {fields["InvoiceId"]} IN ({string.Join(", ", parameterNames)})";

            await using var command = await CreateCommandAsync(commandText, cancellationToken).ConfigureAwait(false);
            for (var i = 0; i < ids.Count; i++)
            {
                AddParameter(command, parameterNames[i], ids[i]);
            }

            var accumulator = new Dictionary<Guid, List<InvoiceLine>>();
            await using var reader = await command.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);
            while (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
            {
                var invoiceId = reader.GetGuid(reader.GetOrdinal("InvoiceId"));
                var description = reader.GetString(reader.GetOrdinal("Description"));
                var quantity = reader.GetDecimal(reader.GetOrdinal("Quantity"));
                var unitPrice = reader.GetDecimal(reader.GetOrdinal("UnitPrice"));
                var tax = reader.GetDecimal(reader.GetOrdinal("Tax"));

                if (!accumulator.TryGetValue(invoiceId, out var list))
                {
                    list = new List<InvoiceLine>();
                    accumulator[invoiceId] = list;
                }

                list.Add(new InvoiceLine(description, quantity, unitPrice, tax));
            }

            return accumulator.ToDictionary(
                pair => pair.Key,
                pair => (IReadOnlyCollection<InvoiceLine>)pair.Value.AsReadOnly());
        }

        private static InvoiceRecord ReadInvoice(DbDataReader reader)
        {
            return new InvoiceRecord(
                reader.GetGuid(reader.GetOrdinal("Id")),
                reader.GetGuid(reader.GetOrdinal("CustomerId")),
                reader.GetGuid(reader.GetOrdinal("VehicleId")),
                reader.GetString(reader.GetOrdinal("Number")),
                reader.GetDateTime(reader.GetOrdinal("Date")),
                reader.GetDecimal(reader.GetOrdinal("Total")),
                reader.GetString(reader.GetOrdinal("Status")));
        }

        private sealed class InvoiceRecord
        {
            public InvoiceRecord(
                Guid id,
                Guid customerId,
                Guid vehicleId,
                string invoiceNumber,
                DateTime invoiceDate,
                decimal totalAmount,
                string status)
            {
                Id = id;
                CustomerId = customerId;
                VehicleId = vehicleId;
                InvoiceNumber = invoiceNumber;
                InvoiceDate = invoiceDate;
                TotalAmount = totalAmount;
                Status = status;
            }

            public Guid Id { get; }

            public Guid CustomerId { get; }

            public Guid VehicleId { get; }

            public string InvoiceNumber { get; }

            public DateTime InvoiceDate { get; }

            public decimal TotalAmount { get; }

            public string Status { get; }
        }
    }
}

File: CRMAdapter/Vast/Adapter/InvoiceAdapter.vb
'-----------------------------------------------------------------------
' File: InvoiceAdapter.vb
' Purpose: Implements canonical invoice access for the VAST Desktop backend.
' Security Considerations: Validates identifiers, clamps limits, parameterizes SQL, and executes through retry/rate-limited policies while sanitizing exception output.
' Example Usage: `Dim invoices = Await adapter.GetByCustomerAsync(customerId, 100, CancellationToken.None)`
'-----------------------------------------------------------------------
Option Strict On
Option Explicit On

Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Data
Imports System.Data.Common
Imports System.Linq
Imports System.Threading
Imports System.Threading.Tasks
Imports CRMAdapter.CommonConfig
Imports CRMAdapter.CommonContracts
Imports CRMAdapter.CommonDomain
Imports CRMAdapter.CommonInfrastructure

Namespace CRMAdapter.Vast.Adapter
    ''' <summary>
    ''' Legacy VAST Desktop implementation of the <see cref="IInvoiceAdapter"/> contract.
    ''' </summary>
    Public NotInheritable Class InvoiceAdapter
        Inherits SqlAdapterBase
        Implements IInvoiceAdapter

        Private Const DefaultListLimit As Integer = 100

        Private Shared ReadOnly RequiredInvoiceKeys As String() = {
            "Invoice.__source",
            "Invoice.Id",
            "Invoice.CustomerId",
            "Invoice.VehicleId",
            "Invoice.Number",
            "Invoice.Date",
            "Invoice.Total",
            "Invoice.Status"
        }

        Private Shared ReadOnly RequiredInvoiceLineKeys As String() = {
            "InvoiceLine.__source",
            "InvoiceLine.InvoiceId",
            "InvoiceLine.Description",
            "InvoiceLine.Quantity",
            "InvoiceLine.UnitPrice",
            "InvoiceLine.Tax"
        }

        Private Shared ReadOnly InvoiceProjectionFields As String() = {
            "Id",
            "CustomerId",
            "VehicleId",
            "Number",
            "Date",
            "Total",
            "Status"
        }

        Private ReadOnly _defaultListLimit As Integer

        ''' <summary>
        ''' Initializes a new instance of the <see cref="InvoiceAdapter"/> class.
        ''' </summary>
        Public Sub New(
            connection As DbConnection,
            fieldMap As FieldMap,
            Optional retryPolicy As ISqlRetryPolicy = Nothing,
            Optional logger As IAdapterLogger = Nothing,
            Optional rateLimiter As IAdapterRateLimiter = Nothing,
            Optional defaultListLimit As Integer = DefaultListLimit)
            MyBase.New(connection, fieldMap, retryPolicy, logger, rateLimiter)
            If defaultListLimit <= 0 Then
                Throw New ArgumentOutOfRangeException(NameOf(defaultListLimit), "Default list limit must be positive.")
            End If

            _defaultListLimit = defaultListLimit
            MappingValidator.EnsureMappings(fieldMap, RequiredInvoiceKeys, NameOf(InvoiceAdapter))
            MappingValidator.EnsureMappings(fieldMap, RequiredInvoiceLineKeys, NameOf(InvoiceAdapter))
            MappingValidator.EnsureEntitySources(fieldMap, New String() {"Invoice", "InvoiceLine"}, NameOf(InvoiceAdapter))
        End Sub

        ''' <inheritdoc />
        Public Function GetByIdAsync(id As Guid, Optional cancellationToken As CancellationToken = Nothing) As Task(Of Invoice) Implements IInvoiceAdapter.GetByIdAsync
            If id = Guid.Empty Then
                Throw New InvalidAdapterRequestException("Invoice id must be provided.")
            End If

            Return ExecuteDbOperationAsync(Of Invoice)(
                "Vast.InvoiceAdapter.GetById",
                Async Function(ct)
                    Dim fieldMap = FieldMap.GetTargets("Invoice", InvoiceProjectionFields)
                    Dim source = FieldMap.GetEntitySource("Invoice")
                    Dim selectClause = String.Join(", ", fieldMap.Select(Function(kvp) $"{kvp.Value} AS [{kvp.Key}]"))
                    Dim commandText = $"SELECT {selectClause} FROM {source} WHERE {fieldMap("Id")} = @id"

                    Dim command = Await CreateCommandAsync(commandText, ct).ConfigureAwait(False)
                    Using command
                        AddParameter(command, "@id", id)
                        Dim reader = Await command.ExecuteReaderAsync(ct).ConfigureAwait(False)
                        Using reader
                            If Not Await reader.ReadAsync(ct).ConfigureAwait(False) Then
                                Return Nothing
                            End If

                            Dim invoiceRecord = ReadInvoice(reader)
                            Dim lines = Await LoadInvoiceLinesAsync(New Guid() {invoiceRecord.Id}, ct).ConfigureAwait(False)
                            Dim lineItems As IReadOnlyCollection(Of InvoiceLine) = Nothing
                            If Not lines.TryGetValue(invoiceRecord.Id, lineItems) Then
                                lineItems = Array.Empty(Of InvoiceLine)()
                            End If

                            Return New Invoice(invoiceRecord.Id, invoiceRecord.CustomerId, invoiceRecord.VehicleId, invoiceRecord.InvoiceNumber, invoiceRecord.InvoiceDate, invoiceRecord.TotalAmount, invoiceRecord.Status, lineItems)
                        End Using
                    End Using
                End Function,
                cancellationToken)
        End Function

        ''' <inheritdoc />
        Public Function GetByCustomerAsync(customerId As Guid, maxResults As Integer, Optional cancellationToken As CancellationToken = Nothing) As Task(Of IReadOnlyCollection(Of Invoice)) Implements IInvoiceAdapter.GetByCustomerAsync
            If customerId = Guid.Empty Then
                Throw New InvalidAdapterRequestException("Customer id must be provided.")
            End If

            Dim limit = EnforceLimit(maxResults, _defaultListLimit, NameOf(maxResults))

            Return ExecuteDbOperationAsync(Of IReadOnlyCollection(Of Invoice))(
                "Vast.InvoiceAdapter.GetByCustomer",
                Async Function(ct)
                    Dim fieldMap = FieldMap.GetTargets("Invoice", InvoiceProjectionFields)
                    Dim source = FieldMap.GetEntitySource("Invoice")
                    Dim selectClause = String.Join(", ", fieldMap.Select(Function(kvp) $"{kvp.Value} AS [{kvp.Key}]"))
                    Dim commandText = $"SELECT TOP (@limit) {selectClause} FROM {source} WHERE {fieldMap("CustomerId")} = @customerId ORDER BY {fieldMap("Date")} DESC"

                    Dim command = Await CreateCommandAsync(commandText, ct).ConfigureAwait(False)
                    Using command
                        AddParameter(command, "@limit", limit, DbType.Int32)
                        AddParameter(command, "@customerId", customerId)

                        Dim invoices As New List(Of InvoiceRecord)()
                        Dim reader = Await command.ExecuteReaderAsync(ct).ConfigureAwait(False)
                        Using reader
                            While Await reader.ReadAsync(ct).ConfigureAwait(False)
                                invoices.Add(ReadInvoice(reader))
                            End While
                        End Using

                        Dim lineLookup = Await LoadInvoiceLinesAsync(invoices.Select(Function(i) i.Id), ct).ConfigureAwait(False)
                        Dim results = invoices.Select(Function(record)
                                                          Dim lineItems As IReadOnlyCollection(Of InvoiceLine) = Nothing
                                                          If Not lineLookup.TryGetValue(record.Id, lineItems) Then
                                                              lineItems = Array.Empty(Of InvoiceLine)()
                                                          End If
                                                          Return New Invoice(record.Id, record.CustomerId, record.VehicleId, record.InvoiceNumber, record.InvoiceDate, record.TotalAmount, record.Status, lineItems)
                                                      End Function).ToList()
                        Return CType(New ReadOnlyCollection(Of Invoice)(results), IReadOnlyCollection(Of Invoice))
                    End Using
                End Function,
                cancellationToken)
        End Function

        Private Function LoadInvoiceLinesAsync(invoiceIds As IEnumerable(Of Guid), cancellationToken As CancellationToken) As Task(Of IReadOnlyDictionary(Of Guid, IReadOnlyCollection(Of InvoiceLine)))
            If invoiceIds Is Nothing Then
                Throw New ArgumentNullException(NameOf(invoiceIds))
            End If

            Return ExecuteDbOperationAsync(Of IReadOnlyDictionary(Of Guid, IReadOnlyCollection(Of InvoiceLine)))(
                "Vast.InvoiceAdapter.LoadLines",
                Async Function(ct)
                    Dim ids = invoiceIds.Distinct().Where(Function(id) id <> Guid.Empty).ToList()
                    If ids.Count = 0 Then
                        Return CType(New Dictionary(Of Guid, IReadOnlyCollection(Of InvoiceLine))(), IReadOnlyDictionary(Of Guid, IReadOnlyCollection(Of InvoiceLine)))
                    End If

                    Dim fields = FieldMap.GetTargets("InvoiceLine", New String() {"InvoiceId", "Description", "Quantity", "UnitPrice", "Tax"})
                    Dim source = FieldMap.GetEntitySource("InvoiceLine")
                    Dim parameterNames = ids.Select(Function(_, index) $"@i{index}").ToArray()
                    Dim selectClause = $"{fields("InvoiceId")} AS [InvoiceId], {fields("Description")} AS [Description], {fields("Quantity")} AS [Quantity], {fields("UnitPrice")} AS [UnitPrice], {fields("Tax")} AS [Tax]"
                    Dim commandText = $"SELECT {selectClause} FROM {source} WHERE {fields("InvoiceId")} IN ({String.Join(", ", parameterNames)})"

                    Dim command = Await CreateCommandAsync(commandText, ct).ConfigureAwait(False)
                    Using command
                        For i = 0 To ids.Count - 1
                            AddParameter(command, parameterNames(i), ids(i))
                        Next

                        Dim accumulator As New Dictionary(Of Guid, List(Of InvoiceLine))()
                        Dim reader = Await command.ExecuteReaderAsync(ct).ConfigureAwait(False)
                        Using reader
                            While Await reader.ReadAsync(ct).ConfigureAwait(False)
                                Dim invoiceId = reader.GetGuid(reader.GetOrdinal("InvoiceId"))
                                Dim description = reader.GetString(reader.GetOrdinal("Description"))
                                Dim quantity = reader.GetDecimal(reader.GetOrdinal("Quantity"))
                                Dim unitPrice = reader.GetDecimal(reader.GetOrdinal("UnitPrice"))
                                Dim tax = reader.GetDecimal(reader.GetOrdinal("Tax"))

                                Dim lines As List(Of InvoiceLine) = Nothing
                                If Not accumulator.TryGetValue(invoiceId, lines) Then
                                    lines = New List(Of InvoiceLine)()
                                    accumulator(invoiceId) = lines
                                End If

                                lines.Add(New InvoiceLine(description, quantity, unitPrice, tax))
                            End While
                        End Using

                        Return CType(accumulator.ToDictionary(Function(pair) pair.Key, Function(pair) CType(pair.Value.AsReadOnly(), IReadOnlyCollection(Of InvoiceLine))), IReadOnlyDictionary(Of Guid, IReadOnlyCollection(Of InvoiceLine)))
                    End Using
                End Function,
                cancellationToken)
        End Function

        Private Shared Function ReadInvoice(reader As DbDataReader) As InvoiceRecord
            Return New InvoiceRecord(
                reader.GetGuid(reader.GetOrdinal("Id")),
                reader.GetGuid(reader.GetOrdinal("CustomerId")),
                reader.GetGuid(reader.GetOrdinal("VehicleId")),
                reader.GetString(reader.GetOrdinal("Number")),
                reader.GetDateTime(reader.GetOrdinal("Date")),
                reader.GetDecimal(reader.GetOrdinal("Total")),
                reader.GetString(reader.GetOrdinal("Status")))
        End Function

        Private NotInheritable Class InvoiceRecord
            Public Sub New(id As Guid, customerId As Guid, vehicleId As Guid, invoiceNumber As String, invoiceDate As DateTime, totalAmount As Decimal, status As String)
                Me.Id = id
                Me.CustomerId = customerId
                Me.VehicleId = vehicleId
                Me.InvoiceNumber = invoiceNumber
                Me.InvoiceDate = invoiceDate
                Me.TotalAmount = totalAmount
                Me.Status = status
            End Sub

            Public ReadOnly Property Id As Guid
            Public ReadOnly Property CustomerId As Guid
            Public ReadOnly Property VehicleId As Guid
            Public ReadOnly Property InvoiceNumber As String
            Public ReadOnly Property InvoiceDate As DateTime
            Public ReadOnly Property TotalAmount As Decimal
            Public ReadOnly Property Status As String
        End Class
    End Class
End Namespace
